Fix some potential memory leaks

firestarman opened the pull request:
fixes #8655

This PR is to fix some potential memory leaks in GPU coalescing operator.

<!--

Thank you for contributing to RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->

firestarman commented:
build

abellina commented:
Nice find. The bug I think is happening because `GpuFilter.apply` used to call `close` the incoming batch but it doesn't anymore: https://github.com/NVIDIA/spark-rapids/pull/7864/files#diff-7718c93b5e463c82611c693aa760c934343e3d8047dd800f1d0c00e9ab8afa52L396

A new method was added "filterAndClose" and I think we may want to do that instead. From me messing about with the code it looks like the only external consumer to this overload of `GpuFilter.apply` is the coalesce class where you found the leak, so can we make this apply private and then use `filterAndClose`?? it will also retry if all the projections are deterministic, and I think that's the intended behavior.

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala:
```diff
@@ -448,11 +448,18 @@ abstract class AbstractGpuCoalesceIterator(
-                  closeOnExcept(GpuFilter(cb, filterExpression)) { filteredCb =>
-                    cb = null // null out `cb` to prevent multiple close calls
+                  // filter the incoming batch if not be filtered.
+                  val filteredCb = if (inputFilterExpression.isEmpty) {
```
Why filter only if the filter isEmpty?

firestarman replied:
When the filter expression is defined, the batch will be filtered at line 424.
```
      var cb = if (inputFilterExpression.isDefined) {
        // If we have reached the cuDF limit once, proactively filter batches
        // after that first limit is reached.
        withResource(cbFromIter)(GpuFilter(_, inputFilterExpression.get))
      } else {
        cbFromIter
      }
```

revans2 reviewed (commented):
That for fixing this. It looks like I caused this when I introduced filterAndClose to GpuFilter and changed some semantics.  Would it be better to just change the calls to GpuFilter.apply to GpuFilter.filterAndClose ?

abellina commented:
BTW I don't think this bug is triggered in 23.04 (when the retry change happened) or 23.06 because the re-hash work made it so we discard the coalesce goal object if the logic in `GpuShuffledHashJoinExec.prepareBuildBatchesForJoin` decides we need a single batch.

Because:

if `needSingleBuildBatch` is true, `sizeOverflow` is false (unless we are in tests, then other cases arise), so we always call `getBuildBatchOptimizedAndClose`, which disregards the filter. 

If `needSingleBuildBatch` is false (`sizeOverflow` must be true, unless we are in tests), then we never call `getAsSingleBatch` which takes into account the `buildGoal`.

So, if I am seeing this right, as of 23.04+ we are not filtering at all. Not until: https://github.com/NVIDIA/spark-rapids/pull/8659 which now takes into account `buildGoal` when `needSingleBuildBatch` is false.

firestarman commented:
build

firestarman commented:
build

firestarman commented:
@abellina @revans2  Could you review it again ?
I have fixed some logic defects for the case when there are 2 or more input batches after entering the filtering mode, and added some unit tests for it.

firestarman commented:
build

firestarman commented:
build

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala:
```diff
@@ -447,72 +457,67 @@ abstract class AbstractGpuCoalesceIterator(
+                      "filtering. Please try increasing your partition count.")
+                    numRows = filteredNumRows
+                    numBytes = filteredBytes
+                  } else {
```
nice catch
